services/src/main/java/com/broadleafcommerce/subscriptionoperation/service/DefaultCancellationPolicyService.java

```diff
@@ -0,0 +1,189 @@
+import com.broadleafcommerce.order.common.domain.enums.subscription.DefaultCancellationPolicySourceTypes;
+import com.broadleafcommerce.order.common.domain.subscription.CancellationPolicyRef;
+import com.broadleafcommerce.rulesengine.expression.service.RuleEvaluationService;
+import com.broadleafcommerce.subscriptionoperation.domain.CancellationPolicy;
+import com.broadleafcommerce.subscriptionoperation.service.provider.CancellationPolicyProvider;
+import com.broadleafcommerce.subscriptionoperation.web.domain.IdentifyCancellationPolicyRequest;
+import com.broadleafcommerce.subscriptionoperation.web.domain.IdentifyCancellationPolicyResponse;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import cz.jirutka.rsql.parser.ast.ComparisonNode;
+import cz.jirutka.rsql.parser.ast.Node;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Default implementation of {@link CancellationPolicyService}.
+ *
+ * @author Sunny Yu
+ * @since 2.x.x
+ */
+@Slf4j
+@RequiredArgsConstructor
+public class DefaultCancellationPolicyService<C extends CancellationPolicy>
+        implements CancellationPolicyService<C> {
+
+    protected static final int DEFAULT_PAGE_SIZE = 50;
+    protected static final String REQUIRES_DIRECT_REFERENCE = "requiresDirectReference";
+
+    @Getter(AccessLevel.PROTECTED)
+    private final CancellationPolicyProvider<C> cancellationPolicyProvider;
+
+    @Getter(AccessLevel.PROTECTED)
+    private final RuleEvaluationService ruleEvaluationService;
+
+    @Getter(AccessLevel.PROTECTED)
+    private final TypeFactory typeFactory;
+
+    @Override
+    public IdentifyCancellationPolicyResponse identifyCancellationPolicy(
+            @lombok.NonNull IdentifyCancellationPolicyRequest request,
+            @Nullable ContextInfo context) {
+        List<CartItem> itemsToCheck = new ArrayList<>(request.getCartItems());
+        IdentifyCancellationPolicyResponse response =
+                typeFactory.get(IdentifyCancellationPolicyResponse.class);
+
+        Page<C> policyPage = null;
+        Node readAllFilters = buildNodeFilters();
+        do {
+            if (CollectionUtils.isEmpty(itemsToCheck)) {
+                return response;
+            }
+
+            Pageable pageable =
+                    (policyPage != null) ? policyPage.nextPageable() : buildFirstPageable();
+            policyPage = cancellationPolicyProvider.readAll(readAllFilters, pageable, context);
+
+            List<C> policies = policyPage.getContent();
+            itemsToCheck.forEach(item -> identifyCancellationPolicy(policies, item)
+                    .ifPresent(policy -> {
+                        response.getPolicyByCartItemId().put(item.getId(),
+                                buildPolicyRef(policy));
+                        itemsToCheck.remove(item);
+                    }));
```

Comment:
You aren't supposed to modify the stream while streaming it. I am not really sure why you are calling `itemsToCheck.remove(item);` anyway.
```suggestion
            itemsToCheck.forEach(item -> identifyCancellationPolicy(policies, item)
                    .ifPresent(policy -> {
                        response.getPolicyByCartItemId().put(item.getId(),
                                buildPolicyRef(policy));
                    }));
```

Reply:
Maybe use a removeIf instead?
```suggestion
        do {
            if (!itemsToCheckIter.hasNext()) {
                return response;
            }

            Pageable pageable =
                    (policyPage != null) ? policyPage.nextPageable() : buildFirstPageable();
            policyPage = cancellationPolicyProvider.readAll(readAllFilters, pageable, context);

            List<C> policies = policyPage.getContent();
            
            itemsToCheck.forEach(item -> identifyCancellationPolicy(policies, item)
                    .ifPresent(policy -> {
                        response.getPolicyByCartItemId().put(item.getId(),
                                buildPolicyRef(policy));
                    }));
            itemsToCheck.removeIf(item -> response.getPolicyByCartItemId().containsKey(item.getId()));
        } while (policyPage.hasNext());
```

Reply:
Ah yeah. Was calling the remove so that it can exit the do while loop to prevent it from keep paging through if the itemsToCheck is empty

Reply:
Or a filter
```suggestion
            itemsToCheck.stream()
            .filter(item -> !response.getPolicyByCartItemId().containsKey(item.getId()))
            .forEach(item -> identifyCancellationPolicy(policies, item)
                    .ifPresent(policy -> {
                        response.getPolicyByCartItemId().put(item.getId(),
                                buildPolicyRef(policy));
                        itemsToCheck.remove(item);
                    }));
```

Reply:
I guess the filter doesn't let you exit early

Reply:
Yeah I'll use the removeIf instead to exit early